torchmultimodal/models/albef.py

```diff
@@ -0,0 +1,244 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import copy
+from collections import namedtuple
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from torchmultimodal.modules.losses.contrastive_loss_with_temperature import (
+    _gather_embeddings_and_labels,
+)
+
+
+ALBEFOutput = namedtuple(
+    "ALBEFOutput",
+    [
+        "image_embeddings",
+        "image_embeddings_m",
+        "text_embeddings",
+        "text_atts",
+        "vl_embeddings",
+        "similarity",
+    ],
+    defaults=(None, None, None, None, None, None),
+)
+
+ALBEFSimilarity = namedtuple(
+    "ALBEFSimilarity",
+    ["sim_i2t", "sim_t2i", "sim_i2t_targets", "sim_t2i_targets"],
+    defaults=(None, None, None, None),
+)
+
+
+class ALBEFModel(nn.Module):
+    def __init__(
+        self,
+        vision_encoder: nn.Module,
+        text_encoder: nn.Module,
+        multimodal_encoder: nn.Module,
+        vision_proj: nn.Module,
+        text_proj: nn.Module,
+        embed_dim: int = 256,
+        queue_size: int = 65536,
+        temp: float = 0.07,
+        momentum: float = 0.995,
+    ):
+        super().__init__()
+        self.vision_encoder = vision_encoder
+        self.text_encoder = text_encoder
+        self.multimodal_encoder = multimodal_encoder
+        self.vision_proj = vision_proj
+        self.text_proj = text_proj
+        self.vision_encoder_m = copy.deepcopy(vision_encoder)
+        self.text_encoder_m = copy.deepcopy(text_encoder)
+        self.multimodal_encoder_m = copy.deepcopy(multimodal_encoder)
+        self.vision_proj_m = copy.deepcopy(vision_proj)
+        self.text_proj_m = copy.deepcopy(text_proj)
+
+        self._copy_params()
+
+        self.queue_size = queue_size
+        self.temp = temp
+        self.momentum = momentum
+
+        self.register_buffer("image_queue", torch.randn(embed_dim, queue_size))
+        self.register_buffer("text_queue", torch.randn(embed_dim, queue_size))
+        self.register_buffer("queue_ptr", torch.zeros(1, dtype=torch.long))
+        self.image_queue = nn.functional.normalize(self.image_queue, dim=0)
```

Comment:
I know the paper uses queues "inspired by MoCo", but could you add some comments explaining why they are needed? also some comments explaining the other class variables above would be great